File: Silver/2017/Jan_p2_hoof_paper_scissors/hps.java
import java.io.*;
import java.util.*;

public class hps {

   private static int get_intv(List<Integer> list, int start, int end) {
      return list.get(end) - list.get(start - 1);
   }

   public static void main(String[] args) throws IOException {

      BufferedReader br = new BufferedReader(new FileReader("hps.in"));

      int num = Integer.parseInt(br.readLine());

      List<Integer> h_list = new ArrayList<>(Collections.singletonList(0));
      List<Integer> p_list = new ArrayList<>(Collections.singletonList(0));
      List<Integer> s_list = new ArrayList<>(Collections.singletonList(0));

      for (int i = 0; i < num; i++) {
         h_list.add(h_list.get(h_list.size() - 1));
         p_list.add(p_list.get(p_list.size() - 1));
         s_list.add(s_list.get(s_list.size() - 1));

         String move = br.readLine();

         if (move.equals("S")) {
               s_list.set(s_list.size() - 1, s_list.get(s_list.size() - 1) + 1);
         } else if (move.equals("H")) {
               h_list.set(h_list.size() - 1, h_list.get(h_list.size() - 1) + 1);
         } else if (move.equals("P")) {
               p_list.set(p_list.size() - 1, p_list.get(p_list.size() - 1) + 1);
         }
      }

      br.close();

      int best_value = Math.max(Math.max(s_list.get(s_list.size() - 1), h_list.get(h_list.size() - 1)), p_list.get(p_list.size() - 1));

      for (int e = 1; e <= num; e++) {
         int curr_value = Math.max(
               Math.max(get_intv(h_list, 1, e), get_intv(s_list, 1, e)),
               get_intv(p_list, 1, e)
         ) + Math.max(
               Math.max(get_intv(h_list, e, num), get_intv(s_list, e, num)),
               get_intv(p_list, e, num)
         );

         if (curr_value > best_value) {
               best_value = curr_value;
         }
      }

      BufferedWriter bw = new BufferedWriter(new FileWriter("hps.out"));
      bw.write(String.valueOf(best_value));
      bw.close();
   }
}

File: Silver/2017/Jan_p2_hoof_paper_scissors/hps.py
f = open("hps.in", "r")
fl = f.read().split("\n")

num = int(fl.pop(0))

h_list = [0]
p_list = [0]
s_list = [0]

for _ in range(num):
   h_list.append(h_list[-1])
   p_list.append(p_list[-1])
   s_list.append(s_list[-1])

   i = fl.pop(0)

   if i == "S":
      s_list[-1] += 1

   elif i == "H":
      h_list[-1] += 1
   
   elif i == "P":
      p_list[-1] += 1

best_value = max(s_list[-1], h_list[-1], p_list[-1])

def get_intv(l,s,e):
   return l[e] - l[s-1]

for e in range(1,num + 1):
   curr_value = max(
      get_intv(h_list,1,e),
      get_intv(s_list,1,e),
      get_intv(p_list,1,e)
   ) + max(
      get_intv(h_list,e,num),
      get_intv(s_list,e,num),
      get_intv(p_list,e,num)
   )

   if curr_value > best_value:
      best_value = curr_value

with open("hps.out", "w") as f:
   f.write(str(best_value))